File: Solutions/C++/21-merge-two-sorted-lists.cpp
/*
Problem Title: 21 - Merge Two Sorted Lists
Lang: C++
Runtime: 12 ms
Memory: 14.9 MB
Problem Url: https://leetcode.com/problems/merge-two-sorted-lists
Submission Url: https://leetcode.com/problems/merge-two-sorted-lists/submissions/764690974
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* head = NULL;
        ListNode* current = NULL;
        
        if(list1 == NULL){
            return list2;
        }
        if(list2 == NULL){
            return list1;
        }
        
        if(list1->val < list2-> val){
            head = list1;
            list1 = list1->next;
        } else {
            head = list2;
            list2 = list2->next;
        }
        current = head;
        
        while(list1 != NULL and list2 != NULL){
            if(list1->val < list2->val){
                current->next = list1;
                list1 = list1->next;
            } else {
                current->next = list2;
                list2 = list2->next;
            }
            current = current->next;
        }
        if(list1 != NULL){
            current->next = list1;
        }
        if(list2 != NULL){
            current->next = list2;
        }
        return head;
    }
};

File: Solutions/C#/Accepted/21-merge-two-sorted-lists.cs
/*
Problem Title: 21 - Merge Two Sorted Lists
Lang: C#
Runtime: 84 ms
Memory: 25.7 MB
Problem Url: https://leetcode.com/problems/merge-two-sorted-lists
Submission Url: https://leetcode.com/problems/merge-two-sorted-lists/submissions/305643080
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode MergeTwoLists(ListNode l1, ListNode l2) {
        
        if(l1 == null){
            return l2;
        }
        if(l2 == null){
            return l1;
        }
        
        ListNode head = null;
        if(l1.val < l2.val){
            head = l1;
            l1 = l1.next;
        }else{
            head = l2;
            l2 = l2.next;
        }
        
        head.next = MergeTwoLists(l1,l2);
        return head;
    }
}